src/metabase/lib/datascript.cljc

```diff
@@ -0,0 +1,1361 @@
+  slated for deletion. For example, from a source to all its columns, from a column to any `:mbql/ref` clause arguments
+  to parent clauses to the join whose condition this is, to all columns on that join, ...
+
+  When new attributes are added to the schema, they should be considered for inclusion here. There's no enforcement, but
+  perhaps we could insist on that at a schema level?"
+  (set (for [[attr {removal :mbql/removal}] schema
+             :when removal]
+         (case removal
+           :mbql.removal/forward attr
+           :mbql.removal/reverse (keyword (namespace attr) (str "_" (name attr)))))))
+
+(def ^:private remove-edge-conditions
+  "Map of edges (forward or backward attributes) to functions which check some larger condition.
+  These condition functions are passed the current set of removals, the source entity, and the target entity.
+  They return a set of removals, possibly with extra things in it.
+
+  Note that the source entity is already in the `removals` set when the function is called."
+  {;; Joins are removed only when their *last* condition is about to be removed. So if we're following a condition->join
+   ;; edge (:mbql.join/_condition) and that puts *all* the conditions on the chopping block, then we add the join.
+   :mbql.join/_condition (fn [removals cnd join]
+                           (cond-> removals
+                             (every? removals (:mbql.join/condition join)) (conj join)))})
+
+(defn- extend-removals
+  "Given a set of things to be removed and an entity from that set, walk any removal edges from that entity and add them
+  to the set. Returns the possibly larger removal set."
+  [removals entity]
+  (into removals
+        (for [edge  remove-edges                             ; For each removal edge:
+              :let [pred    (remove-edge-conditions edge)
+                    targets (when-let [t (get entity edge)]
+                              (if (set? t) t #{t}))]
+              :when targets                                  ; If that edge exists on this entity
+              target targets                                 ; For each target of the edge
+              :when (or (not pred)                           ; When either there's no predicate for this edge
+                        (pred removals entity target))]      ; or the predicate passes
+          target)))                                          ; Add that target to the set.
+
+(defn- removal-closure
+  "Recursively expands a set of removals until it stops growing."
+  [removals]
+  (loop [removals removals]
+    (let [removals' (reduce extend-removals removals removals)]
+      (if (= (count removals) (count removals'))
+        removals'
+        (recur removals')))))
+
+(defn- removals->tx [removals]
+  (vec (for [removed removals]
+         [:db.fn/retractEntity (:db/id removed)])))
+
+(defn- fix-series-indexing
```

Comment:
Could we skip this if we appended by `(inc greatest-series-value)` instead of `(count series-members)`? I imagine we would introduce "holes" in the numbering, but would that break thing?

Reply:
I'm moving to this model - you're exactly right.